src/gtest/test_proofs.cpp

```diff
@@ -0,0 +1,511 @@
+#include <gtest/gtest.h>
+#include "zcash/Proof.hpp"
+
+#include <iostream>
+
+#include "libsnark/common/default_types/r1cs_ppzksnark_pp.hpp"
+#include "libsnark/zk_proof_systems/ppzksnark/r1cs_ppzksnark/r1cs_ppzksnark.hpp"
+#include "zk_proof_systems/ppzksnark/r1cs_ppzksnark/r1cs_ppzksnark.hpp"
+#include "relations/constraint_satisfaction_problems/r1cs/examples/r1cs_examples.hpp"
+
+using namespace libzcash;
+
+typedef libsnark::default_r1cs_ppzksnark_pp curve_pp;
+typedef libsnark::default_r1cs_ppzksnark_pp::G1_type curve_G1;
+typedef libsnark::default_r1cs_ppzksnark_pp::G2_type curve_G2;
+typedef libsnark::default_r1cs_ppzksnark_pp::GT_type curve_GT;
+typedef libsnark::default_r1cs_ppzksnark_pp::Fp_type curve_Fr;
+typedef libsnark::default_r1cs_ppzksnark_pp::Fq_type curve_Fq;
+typedef libsnark::default_r1cs_ppzksnark_pp::Fqe_type curve_Fq2;
+
+#include "streams.h"
+#include "version.h"
+#include "utilstrencodings.h"
+
+TEST(proofs, sqrt_fq)
+{
+    // Poor man's PRNG
+    curve_Fq acc = curve_Fq("348957923485290374852379485") ^ 1000;
+
+    size_t quadratic_residues = 0;
+    size_t quadratic_nonresidues = 0;
+
+    for (size_t i = 1; i < 1000; i++) {
+        try {
+            acc += curve_Fq("45634563456") ^ i;
+
+            curve_Fq x = acc.sqrt();
+            ASSERT_TRUE((x*x) == acc);
+            quadratic_residues += 1;
+        } catch (std::runtime_error &e) {
+            quadratic_nonresidues += 1;
+        }
+    }
+
+    // Half of all nonzero elements in Fp are quadratic residues
+    ASSERT_TRUE(quadratic_residues == 511);
+    ASSERT_TRUE(quadratic_nonresidues == 488);
```

Comment:
Ran the test and hard-coding. Half of the nonzero elements in Fp are quadratic residues, so we should expect approximately half of the random elements we sample to be such. I wanted the test to be deterministic because otherwise we have to write some kind of standard deviation logic.